docs/Style.thy

```diff
@@ -0,0 +1,182 @@
+   limit. Note also that the \equiv symbol is used.
+ *)
+
+definition hd_opt :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt \<equiv> case_list None (\<lambda>h t. Some h)"
+
+(* In the case that the type_signature causes an overflow of the 100 char line limit,
+   use the following pattern.
+
+   Note that the indentation of the type signature remains no less than the indentation level of
+   the initial `"`.
+ *)
+
+type_synonym my_very_long_nat_type = nat
+type_synonym 'a my_very_long_option_type = "'a option"
+type_synonym 'a my_very_long_list_type = "'a list"
+definition hd_opt2 :: "'a my_very_long_list_type \<Rightarrow> my_very_long_nat_type \<Rightarrow>
+                       'a my_very_long_list_type \<Rightarrow> 'a my_very_long_option_type" where
+  "hd_opt2 dummy_list dummy_nat \<equiv> case_list None (\<lambda>h t. Some h)"
+
+(* In the case that the definition body causes an overflow of the 100 char line limit, use one
+   of the following two patterns.
+
+   Note that in each case, indentation of the right-hand expression remains no less than the
+   indentation level of `case_list`.
+ *)
+
+definition hd_opt3 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt3 \<equiv> case_list
+             None
+             (\<lambda>h t. Some h)"
+
+definition hd_opt4 :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt4 \<equiv>
+   case_list None (\<lambda>h t. Some h)"
+
+
+section \<open>Lemma Statements\<close>
+
+(* Lemmas should be written over 2 lines in the following style.
+
+   Note that indentation is 2 spaces and the lemma statement does not run past 100 character line
+   limit. Note also that multiple premises are listed within `\<lbrakk> \<rbrakk>`.
+
+  For example,
+*)
+
+lemma test_lemma0:
+  "\<lbrakk>hd_opt l \<noteq> None; hd_opt (hd l) = Some x\<rbrakk> \<Longrightarrow>
+   hd_opt (concat l) = Some x"
+  apply (clarsimp simp: hd_opt_def
+                 split: list.splits)
+  done
+
+section \<open>Proofs\<close>
+
+subsection \<open>Line length: 100\<close>
+
+(* The line length should be restricted to 100 characters wherever possible. See below for how to
+   indent statements that need to be broken over multiple lines. *)
+
+subsection \<open>Indentation (general)\<close>
+
+(* From Gerwin's guide:
+
+   If an apply command is applied to a state with n+1 subgoals, it must be indented by n spaces
+   relative to the first apply in the sequence (see the Isabelle distribution for many examples).
+   This indicates textually which line solved a subgoal and how many new subgoals where generated.
+   Invaluable information when something goes wrong because something changed somewhere. Also, it
+   gives a nice visual indication of the nesting depth of your proof.
+
+   The usual indentation for 0 subgoals is 2 spaces for a top-level proof.
+
+  For example:
+*)
+
+lemma test_lemma1:
+  "2 \<le> length l \<Longrightarrow> \<exists>h1 h2 t. l = h1 # h2 # t"
+  apply (case_tac l)
```

Comment:
Should probably have `cases` in the example since we should prefer that where possible.